Create test WE for Bug 1292234

biancadanforth opened the pull request:
See ./bug-1292234-test-we/README.md for context

See also [WIP Bugzilla patch](https://github.com/biancadanforth/bug_1292234)

biancadanforth committed:
Create test WE for Bug 1292234

rhelmer commented:
One meta-comment I should have done earlier - I don't think the mdn/webextensions-examples repo is the right place for this, since it'll cause confusion for regular third-party extension devs.

I think https://docs.telemetry.mozilla.org/ is probably the right place since it's pretty specific to Mozilla's experiments program (for instance the client guidelines https://docs.telemetry.mozilla.org/cookbooks/client_guidelines.html).

*Maybe* http://firefox-source-docs.mozilla.org/ but I feel like that's kind of stretch, I think the above is probably a better place for the documentation side of this.

The actual demo extension should live somewhere too but that could just be a GH repo in the mozilla/ namespace IMHO shouldn't be the generic MDN webextension one.

rhelmer commented on bug-1292234-test-we/background.js:
```diff
@@ -0,0 +1,15 @@
+    console.log("WebExtensionLog::ExtensionLocalStorageChange", changes);
+  }
+  
+  // Add listeners
```
I don't think this comment adds clarity, since it's just describing what is happening on the next line.

If you want to document *why* adding a handler is a good idea here, that would be worth commenting.

biancadanforth replied:
Good point thanks

rhelmer commented on bug-1292234-test-we/web-ext-config.js:
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  run: {
+    // Points to my local artifact development build of Firefox
+    firefox: '/Users/bdanforth/src/mozilla-unified/objdir-frontend-debug-artifact/dist/Nightly.app/Contents/MacOS/firefox',
```
This seem pretty specific for your setup :)

biancadanforth replied:
Is that bad? How might I abstract that away? One way or the other, I need to point to my local build.

rhelmer commented on bug-1292234-test-we/web-ext-config.js:
```diff
@@ -0,0 +1,9 @@
+    // Points to my local artifact development build of Firefox
+    firefox: '/Users/bdanforth/src/mozilla-unified/objdir-frontend-debug-artifact/dist/Nightly.app/Contents/MacOS/firefox',
+    startUrl: [
+      'about:debugging',
```
I didn't know `web-ext` supported this that's super cool.

rhelmer reviewed (approved):
Just nits take or leave it.

biancadanforth commented:
@rhelmer Thanks for your [feedback](https://github.com/biancadanforth/webextensions-examples/pull/3#issuecomment-468818919).

> One meta-comment I should have done earlier - I don't think the mdn/webextensions-examples repo is the right place for this, since it'll cause confusion for regular third-party extension devs.

This is my own personal fork of the repo, I am not opening PRs onto the upstream MDN repo, nor am I using this repo for any other reason than my own personal experimentation and WE prototyping. Isn't that one of the use cases of forks -- to extend some code for your own purposes?

I agree that I have only added a few of my own example extensions here, and they could probably all live in a separate repo from this fork.

> I think https://docs.telemetry.mozilla.org/ is probably the right place since it's pretty specific to Mozilla's experiments program (for instance the client guidelines https://docs.telemetry.mozilla.org/cookbooks/client_guidelines.html).

I'm a little confused what part of this would go into official Firefox docs?

These extensions are not intended to be examples for a larger audience; just as a place for me to try things out myself. I do think if/when a fix for Bug 1292234 lands, we could add a section to the doc you created about preferred use of preferences in experiments and link to an example extension hosted somewhere more official.

rhelmer commented:
OK good point about just using this in your personal repo :) Since it was an MDN webextensions-examples fork I assumed you intended to ultimately land it in the upstream repo and was suggesting other places this could go potentially but you're right that this isn't really necessary. Sorry for the confusion.

biancadanforth committed:
Put something in window.indexedDB storage for testing

In ./devtools/shared/specs/storage.js, the `storeObjectType` provides a schema of sorts our [WIP patch](https://gist.github.com/biancadanforth/fb6aaae07084512a594a8098c971807e), which currently uses an existing `storeObjectType` of `"storagestoreobject"`. This is the `storeObjectType` used by the window.localStorage and window.sessionStorage actors.

However, this `"storeagestoreobject"` type [limits the data types](https://searchfox.org/mozilla-central/source/devtools/shared/specs/storage.js#119-122) of values stored to strings. Extension storage like `browser.storage.local` have fewer restrictions on the types of values that can be stored compared to window.localStorage and window.sessionStorage.

Extension storage key-value type restrictions are most similar to (but more restrictive than[1]) `window.indexedDB`, so this patch puts some items in IndexedDB to compare how the IndexedDB storage actor handles and processes the data as a basis to create a new `storeObjectType` for extension storage.

[1]:
Compared to [`window.localStorage` and `window.sessionStorage`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage):
> The values stored can be any JSON-ifiable value, not just String.
Compared to [`window.indexedDB`](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API#Key_concepts_and_usage):
While extension storage can store any JSON-ifiable value, IndexedDB can store:
> any objects supported by the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/Guide/API/DOM/The_structured_clone_algorithm)
This includes Maps, for example and other complex objects.

biancadanforth committed:
Add browserAction popup

This is to see what happens when there is no background page (i.e. by removing the 'background' key from 'manifest.json') for an extension (say it just has a popup) when I open the Storage inspector.

Right now we are using the 'browser.storage' WE API directly off of the extension's 'window' object (e.g.:

```
  // TODO: Don't use WE APIs directly, as a WE can overwrite the `browser` object
  async removeItem(host, name) {
    const win = this.storageActor.getWindowFromHost(host);
    await Cu.waiveXrays(win).browser.storage.local.remove(name);
  },
```

...but `window` not always available as in the case when there is no background page running in the extension and no other extension pages are open. I have confirmed this does indeed break with the following error:

```
Protocol error (unknownError): Error occurred while creating actor' server2.conn0.child13/storageActor8: TypeError: Cu.waiveXrays(...).browser is undefined
```

...in ui.js:165:15. So we need to rely on something other than 'window' to get at the extension's storage.

It may be possible if we can get the addonId some way to access the extension's storage via the privileged implementation of the storage WE API.

biancadanforth committed:
Add extension page to use Browser Content Toolbox debugger

Because the storage actor (in mozilla-central/devtools/server/actors/storage.js) for Bug 1292234 is running in the extension child process, it cannot be debugged from the Browser Toolbox (debugger for main process), the addon debugger (debugger for web extension itself) or the in-content toolbox (debugger for web page scripts).

With Fission, there will be a single toolbox that can target any different process, but in the meantime, there is a workaround to be able to debug this storage actor with the Browser Content Toolbox.

1) Open a webextension page of the target extension in a Firefox tab
2) Open the “Browser Content Toolbox” (from App Menu -> “Web Developers”) while the webextension page tab is currently selected (this way a toolbox window connected to the extension child process should be opened)

biancadanforth committed:
Open an extension page in popup

This helps debug the situation where an extension does not have a background script (and may therefore not have a window object available to our storage actor.
